### PR TITLE
Introduces Telemetry to the README

### LINKED ISSUE #826: Update ReadMe with Telemetry usage information
Please update ReadMe with Telemetry usage information

### PR DESCRIPTION
**Pull request by iambmelt**
First pass at the README section for introducing telemetry.

Closes #826

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -415,7 +415,34 @@ Logger.getInstance().setLogLevel(Logger.LogLevel.Verbose);
   adb logcat > "C:\logmsg\logfile.txt"
  ```
  More examples about adb cmds: https://developer.android.com/tools/debugging/debugging-log.html#startingLogcat
- 
+
+#### Telemetry
+
+ADAL provides a built-in callback mechanism to supply consuming applications with event data generated during requests. The event data is sanitized of Personally Identifiable Information (PII) and Organizationally Indentifiable Information (OII) and is designed to give consumers of the library insight into the performance, reliability, and usage of ADAL.
```
Comment (heidijinxujia) [inline README.md]: Shouldn't it be Organizationally Identifiable Information?

Comment (iambmelt) [inline README.md]: Oops. Spelling.

Comment (iambmelt) [inline README.md]: Resolved 👍

File: README.md [hunk 2]
```diff
@@ -415,7 +415,34 @@ Logger.getInstance().setLogLevel(Logger.LogLevel.Verbose);
   adb logcat > "C:\logmsg\logfile.txt"
  ```
  More examples about adb cmds: https://developer.android.com/tools/debugging/debugging-log.html#startingLogcat
- 
+
+#### Telemetry
```
Comment (weijjia) [inline README.md]: @iambmelt Daniel and I talked about that for MSAL, the documentation on diagnostics part should be in wiki, we can link the wiki on readme. I'm not sure if we want to do the same thing for ADAL, double check with him. If we do it in wiki, you can expand more on the default dispatcher and the aggregated one without overwhelming readme :)

File: README.md [hunk 3]
```diff
@@ -415,7 +415,34 @@ Logger.getInstance().setLogLevel(Logger.LogLevel.Verbose);
...
+
+ADAL provides a built-in callback mechanism to supply consuming applications with event data generated during requests. The event data is sanitized of Personally Identifiable Information (PII) and Organizationally Identifiable Information (OII) and is designed to give consumers of the library insight into the performance, reliability, and usage of ADAL.
```
Comment (nazukj) [inline README.md]: >event data [](start = 82, length = 10)

i would recommend saying event data (telemetry)

File: README.md [hunk 4]
```diff
@@ -415,7 +415,34 @@ Logger.getInstance().setLogLevel(Logger.LogLevel.Verbose);
...
+
+To capture telemetry data, add the following to your application subclass:
+```java
+// Get a reference to the Telemetry singleton
+private static final Telemetry sTelemetry = Telemetry.getInstance();
+
+// Flag to turn event aggregation on/off
+private static final boolean sTelemetryAggregationIsRequired = false;
```
Comment (nazukj) [inline README.md]: >sTelemetryAggregationIsRequired [](start = 29, length = 31)

can you please add more information explaining aggregation and what true vs false does with examples?

File: README.md [hunk 5]
```diff
@@ -415,7 +415,34 @@ Logger.getInstance().setLogLevel(Logger.LogLevel.Verbose);
...
+    sTelemetry.registerDispatcher(new IDispatcher() {
+        @Override
+        public void dispatchEvent(Map<String, String> events) {
+            // Events from ADAL will be sent to this callback
+
+            // If sTelemetryAggregationIsRequired == true, all events generated per-request
+            // will be aggregated into a single event and published once.
+
+            // If false, events will be published one-at-a-time throughout the various phases
+            // of a an authentication request.
```
Comment (nazukj) [inline README.md]: can we possibly add example of what the data would look like in either case to help clarify. I would also add a little more information on pros and cons of each and when a developer should choose one vs the other. Doesnt have to all be in readme. happy to link off to another document with very detailed information but definitively see the value in having this information handy for the dev

File: README.md (hunk 2 above)
Comment (danieldobalian) [inline README.md]: +1 on Wiki over README (even for ADAL).  A snapshot of telemetry may make sense, but if we want to show the schema, more advanced usage (sending it to a service), and more detail overall the Wiki is the right spot.

File: README.md (hunk 4 above)
Comment (addev-ashish) [inline README.md]: we are asking people to keep this as true

Comment (addev-ashish) [inline README.md]: also you might want to add what aggregation does

File: README.md [hunk 6]
```diff
@@ -415,7 +415,34 @@ Logger.getInstance().setLogLevel(Logger.LogLevel.Verbose);
...
+
+// Flag to turn event aggregation on/off
+private static final boolean sTelemetryAggregationIsRequired = false;
+static {
+    sTelemetry.registerDispatcher(new IDispatcher() {
+        @Override
+        public void dispatchEvent(Map<String, String> events) {
+            // Events from ADAL will be sent to this callback
+
+            // If sTelemetryAggregationIsRequired == true, all events generated per-request
```
Comment (addev-ashish) [inline README.md]: say that true is reccomended, unless teams want to aggregate by themselves.

File: README.md (hunk 3 above)
Comment (iambmelt) [inline README.md]: Resolved 👍

File: README.md (hunk 5 above)
Comment (iambmelt) [inline README.md]: Will do. Will break out into a wiki page and link from README 👍

**Comment by iambmelt** [comment]
To Do:
- Break telemetry docs out into wiki page
- Include sample event data for aggregated vs non-aggregated

**Comment by iambmelt** [comment]
Closing - per comments from @weijjia & @danieldobalian Telemetry info will be provided [in the Wiki](https://github.com/AzureAD/azure-activedirectory-library-for-android/wiki/Telemetry)

**Comment by iambmelt** [comment]
Telemetry Wiki has been updated with sample Events.

[Link to Wiki.](https://github.com/AzureAD/azure-activedirectory-library-for-android/wiki/Telemetry)

**Comment by iambmelt** [comment]
Reopening PR to add hyperlink to README (to wiki)

**Comment by iambmelt** [comment]
README/Wiki have been updated per discussion in scrum:
- README contains brief introduction to the concept and purpose of telemetry + hyperlink to [Wiki:Telemetry](https://github.com/AzureAD/azure-activedirectory-library-for-android/wiki/Telemetry)
- Wiki has been updated with new introduction (which is not duplicative of README paragraph) + sample aggregated/non-aggregated event data in JSON format

/cc reviewers: @weijjia @danieldobalian @nazukj @addev-ashish @heidijinxujia

**Pull request merged**